### DISCUSSION TITLE
PSA: Deleting collection does not delete _externalAuths

### DISCUSSION THREAD

**Discussion by pnmcosta** [pocketbase/pocketbase#2570] [Q&A]
Happened as I dropped the collection cause I did significant changes to it and migration would recreate it (it was quicker this way then to create a specific migration)

I dropped the collection on the testing instance and deploy when smooth, unfortunately I couldn't re-login with the same auth account keep getting a 400 with a "sql no rows found" error on the `/auth-with-oauth2` enpoint, downloading a backup of the db confirms that there are dangling `_externalAuths`, e.g. when the collection was dropped this table was not pruned.

I'm sure a fix will be done soon, in the meantime, avoid dropping collection auth's without deleting records first ;)

**Comment by ganigeorgiev** [comment]
Hm, that's strange because there is a foreign key with cascade constraint on the `_externalAuths.collectionId` column and the related entries should have been deleted automatically. I'll investigate it sometime later this week.

**Reply by ganigeorgiev** [reply]
It seems that I didn't initialized the `foreign_keys` PRAGMA correctly. In https://github.com/pocketbase/pocketbase/blob/master/core/db.go#L16 should be `ON`, not `TRUE`. I'll update it a little bit later and will add a test.

**Reply by pnmcosta** [reply]
Can we expect our record relations to cascade deletes too once this is done? Just asking cause I do have a task to handle that manually via event hooks for my custom relations.

**Reply by ganigeorgiev** [reply]
@pnmcosta Actually I'm not able to reproduce it. Although `ON` is used in the SQLite examples, the [value is marked as `boolean`](https://www.sqlite.org/pragma.html#pragma_foreign_key_list) so `TRUE` should also work.

How did you delete the collection?

---

> Can we expect our record relations to cascade deletes too once this is done?

There are no foreign keys for the records relations.

**Reply by pnmcosta** [reply]
> How did you delete the collection?

Just dropped it on the admin UI:

![image](https://github.com/pocketbase/pocketbase/assets/550684/474de64c-728b-458b-bcc5-e73fdbd5b6a8)


> There are no foreign keys for the records relations.

Cool

**Reply by ganigeorgiev** [reply]
What version of PocketBase do you use? Are you using the prebuilt executable?

**Reply by pnmcosta** [reply]
My fork https://github.com/pnmcosta/pocketbase I need the Instragram provider, I'm replacing it in go mod:

`replace github.com/pocketbase/pocketbase v0.16.2 => github.com/pnmcosta/pocketbase v0.0.0-20230524202557-5a6a64e3cc60`

I am building with `CGO_ENABLED=1`

**Reply by pnmcosta** [reply]
oh wait might be important too, I'm using https://litestream.io/ to replicate the DB where that happened.

**Reply by ganigeorgiev** [reply]
I've pushed in `master` to use `ON` just for consistency with the SQLite3 examples and added a test on collection delete, but I'm not able to reproduce it locally with both `ON/TRUE` and `CGO_ENABLED=0/1`.
It is possible that something with the replication has gone wrong, I'm not sure.

**Comment by pnmcosta** [comment]
I was able to replicate it locally without litestream, let me update my fork with that change and try again, I'll keep db backups and share privately if you'd like.

**Reply by pnmcosta** [reply]
Ok once I've updated the fork and used it in a build it worked! No dangling external auths. Many thanks.

**Reply by ganigeorgiev** [reply]
But I don't think the change has any real effect. I'm still not able to reproduce the issue locally.

If you have time, you can try to modify https://github.com/pnmcosta/pocketbase/blob/master/core/db.go#L16 to: `ON`, `TRUE`, `FALSE`, and test each option by running `go test ./daos/... -run="TestDeleteCollection" -v` (only the last one should fail).

**Reply by pnmcosta** [reply]
Yup confirmed, those tests are positive on both `ON` or `TRUE` in my system too, it's strange it did happen on my fly.io instance and locally earlier. I'm not too worried about it, don't expect to be dropping collections too often 🤣

**Comment by ganigeorgiev** [comment]
@pnmcosta Another user reported similar issue and found the culprit in https://github.com/pocketbase/dbx/issues/3. I've submitted a fix [v0.16.3 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.16.3).

**Reply by pnmcosta** [reply]
Brilliant, thank you for persisting on it. I'll rebase my fork and get it up to date.
